### PR TITLE
WEB-1414 ESLint Update

### PR DESCRIPTION
**Pull request by mariavargas831**
This PR updates ESLint version and related plugins

- ESLint has been updated to version 7.32.0 from 5.16.0
- Rules for ESLint and related plugins have been added/updated
- The latest ESLint version is 8.0.0 but compat, jsx-a11y and react eslint plugins are not yet compatible with 8.0.0, so I decided to go with 7.32.0 for now
- babel-eslint was deprecated and moved to @babel/eslint-parser
- Node and Commonjs rules were moved into a separate plugin - eslint-plugin-node, this was installed and new rules added
- This is still a work in progress but can start being reviewed

### CODE REVIEW

File: src/plugin/filenames.js [hunk 1]
```diff
@@ -5,7 +5,7 @@
  *
  * Not overly strict, but ensure you're always lower case, and kebab.
  */
-module.exports = {
+ module.exports = {
```
Comment (mingteowork) [inline src/plugin/filenames.js]: Do we have linting rules for the linting files? 🤣

File: src/plugin/import--node.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 /**
- * Import Syntax Rules (v2.17.2)
+ * Import Syntax Rules (v2.25.2)
  *
  * https://github.com/benmosher/eslint-plugin-import
  */
-module.exports = {
+ module.exports = {
```
Comment (mingteowork) [inline src/plugin/import--node.js]: It's here too!

File: src/plugin/react.js [hunk 3]
```diff
@@ -74,25 +81,30 @@ module.exports = {
...
         'react/jsx-no-bind': 'off',                                              // Prevent usage of .bind() and arrow functions in JSX props
         'react/jsx-no-comment-textnodes': 'error',                               // Prevent comments from being inserted as text nodes
+        'react/jsx-no-constructed-context-values': 'error',                      // Prevent react contexts from taking non-stable values
         'react/jsx-no-duplicate-props': ['error', {ignoreCase: true}],           // Prevent duplicate props in JSX
         'react/jsx-no-literals': 'off',                                          // Prevent usage of unwrapped JSX strings
+        'react/jsx-no-script-url': 'error',                                      // Prevent usage of javascript: URLs
         'react/jsx-no-target-blank': 'error',                                    // Prevent usage of unsafe target='_blank'
         'react/jsx-no-undef': 'error',                                           // Disallow undeclared variables in JSX
+        'react/jsx-no-useless-fragment': 'error',                                // Disallow unnecessary fragments
```
Comment (mingteowork) [inline src/plugin/react.js:106]: This is a nice one 👍🏻

**Review by karunabaral — APPROVED** [review]

File: src/plugin/node.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        'node/prefer-global/buffer': ['error', 'always'],            // enforce either Buffer or require("buffer").Buffer
+        'node/prefer-global/console': ['error', 'always'],           // enforce either console or require("console")
+        'node/prefer-global/process': ['error', 'always'],           // enforce either process or require("process")
+        'node/prefer-global/text-decoder': ['error', 'always'],      // enforce either text-decoder or require("util").TextDecoder
+        'node/prefer-global/text-encoder': ['error', 'always'],      // enforce either text-encoder or require("util").TextEncoder
+        'node/prefer-global/url-search-params': ['error', 'always'], // enforce either URLSearchParams or require("url").URLSearchParams
+        'node/prefer-global/url': ['error', 'always'],               // enforce either URL or require("url").URL
+        'node/prefer-promises/dns': 'error',                         // disallow callback API in favor of promise API
+        'node/prefer-promises/fs': 'error',                          // disallow callback API in favor of promise API
+    },
```
Comment (francox9) [inline src/plugin/node.js:56]: One of the major issues we want to solve with eslint update is `Stop linter complaining about optional chaining function execution, that doesn’t return, like  someObject?.someFunc?.()` (as described in the ticket) Is that already addressed (or found can't be done)?

Comment (francox9) [reply src/plugin/node.js:56]: For context, atm for code like below we have to suppress the lint error, or do `void event?.preventDefault?.();`

```
export const preventDefaultThenAction = (actionName) => {
    const thenAction = action(`event: ${actionName}`);

    return (event) => {
        event?.preventDefault?.(); // eslint-disable-line no-unused-expressions

        thenAction();
    };
};
```

File: src/plugin/react.js [hunk 5]
```diff
@@ -24,10 +24,13 @@ module.exports = {
...
         'react/forbid-dom-props': 'off',                             // Forbid certain props on DOM Nodes
         'react/forbid-elements': 'off',                              // Forbid certain elements
-        'react/forbid-prop-types': ['error', {forbid: ['array']}],   // Forbid certain propTypes
         'react/forbid-foreign-prop-types': 'error',                  // Forbid using another component’s prop types unless specifically imported/exported
+        'react/forbid-prop-types': ['error', {forbid: ['array']}],   // Forbid certain propTypes
+        'react/function-component-definition': 'off',                // Enforce a specific function type for function components
         'react/no-access-state-in-setstate': 'error',                // Prevent using this.state within a this.setState
+        'react/no-adjacent-inline-elements': 'off',                  // Prevent adjacent inline elements not separated by whitespace
         'react/no-array-index-key': 'off',                           // Prevent usage of Array index in keys
+        'react/no-arrow-function-lifecycle': 'error',                // Prevent usage of arrow functions for lifecycle methods
```
Comment (francox9) [inline src/plugin/react.js]: lol isn't this where i got commented 100 times

File: package.json [hunk 6]
```diff
@@ -8,22 +8,28 @@
     "eslint-plugin"
   ],
   "engines": {
-    "node": ">=0.10.0"
+    "node": "^12.22.0 || ^14.17.0 || >=16.0.0"
```
Comment (seankoji) [inline package.json]: should we just say `>=12.22.0`?

File: src/plugin/import--browserify.js [hunk 7]
```diff
@@ -22,15 +22,16 @@ module.exports = {
         'import/no-cycle': 'off',                                               // Ensures that there is no resolvable path back to this module via its dependencies.
         'import/no-useless-path-segments': ['error', {noUselessIndex: true}],   // Use this rule to prevent unnecessary path segments in import and require statements
         'import/no-relative-parent-imports': 'off',                             // Use this rule to prevent imports to folders in relative parent paths
-        'import/no-unused-modules': ['error', {missingExports: true}],          // Report modules without exports, or exports without matching import in another module
+        'import/no-relative-packages': 'off',                                   // Use this rule to prevent import packages through relative paths
```
Comment (seankoji) [inline src/plugin/import--browserify.js:25]: whats the thinking behind turning this off?

File: src/plugin/node.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/**
+ * v11.1.0
+ * Additional ESLint rules for Node.js
+ *
+ * https://github.com/mysticatea/eslint-plugin-node
+ */
+ module.exports = {
+    plugins: ['node'],
+    rules: {
```
Comment (seankoji) [inline src/plugin/node.js:9]: which of these, if any, are new? or were they just moved from another file

**Comment by seankoji** [comment]
should we add this one in too while you're here? with both rules set to error i would say 

 https://www.npmjs.com/package/eslint-plugin-react-hooks 

it will introduce a few code changes, but i think its worthwhile

**Review by seankoji — APPROVED** [review]

File: src/plugin/node.js (hunk 4 above)
Comment (seankoji) [reply src/plugin/node.js:56]: I think that may have been resolved in [eslint 7.5](https://eslint.org/blog/2020/07/eslint-v7.5.0-released#optional-chaining-support) 

> Support for optional chaining has landed in ESLint!

so hopefully that means it'll go away automatically! perhaps give it a test @francox9

**Review by karunabaral — APPROVED** [review]

File: src/plugin/node.js (hunk 8 above)
Comment (mariavargas831) [reply src/plugin/node.js:9]: There were quite a few new ones, so I've grouped them together for now so it's easier to review. All new rules are from line 22 onward.

File: src/plugin/node.js (hunk 4 above)
Comment (mariavargas831) [reply src/plugin/node.js:56]: yeah, this has been resolved by eslint, I no longer get that error when testing

**Review by karunabaral — APPROVED** [review]

File: src/rules/stylistic-issues.js [hunk 9]
```diff
@@ -13,7 +13,7 @@ module.exports = {
         'camelcase': ['error', {                                                                  // enforce camelcase naming convention
             properties: 'never',
             ignoreDestructuring: true,
-            allow: ['^UNSAFE_'],
+            allow: ['^UNSAFE_', '^[a-zA-Z0-9-]+_{2}[a-zA-Z0-9-]+$'],
```
Comment (mariavargas831) [inline src/rules/stylistic-issues.js]: this regex is to allow JSX naming like SomeComponent__SubComponent

Comment (seankoji) [reply src/rules/stylistic-issues.js]: put that comment inline next to the rule

**Review by nesamouzehkesh — APPROVED** [review]

File: package.json [hunk 10]
```diff
@@ -8,22 +8,30 @@
...
   "engines": {
-    "node": ">=0.10.0"
+    "node": ">=12.22.0"
   },
   "license": "ISC",
-  "author": "Kayo & Fox Sports Web Developers",
+  "author": "Streamotion Web Developers",
+  "repository": {
+    "type": "git",
+    "url": "ssh://git@github.com:FoxSportsAustralia/eslint-config-foxsports.git"
```
Comment (karunabaral) [inline package.json:17]: You will need to move this repo to FSA-Streamotion space to publish. Chat with Nesa/Szasza about it.

**Review by karunabaral — APPROVED** [review]

**Review by seankoji — APPROVED** [review]

**Pull request merged**